PITCHME.md

```diff
@@ -0,0 +1,530 @@
+
++++
+
+  C++
+-----
+
+- Build : GNU make, CMake, QMake, Ninja, Gradle, SCons, Premake, etc.
+- Docs  : Doxygen, Sphinx, etc.
+- Tests : Boost.Test, cppunit, CxxTest, gtest, etc.
+- Logs  : Boost.Log, log4cpp, log4cxx, etc.
+- Style : astyle, uncrustify, etc.
+- Analysis: cppcheck, cpplint, clang, etc.
+- packaging: deb, rpm, CPack, etc. 
+
++++
+
+Python
+------
+
+- Build : Python
+- Docs  : Sphinx
+- Tests : unitest, doctest, nose, pytest
+- Logs  : logging
+- Style : PEP8
+- Analysis: mypy
+- Packaging: pip + setuptools
+
++++
+
+Static
+------
+
+- Check what you can
+- Compile what you need
+- Don't change
+- Be fast
+
++++
+
+Dynamic
+-------
+
+- Check if you can
+- Compile if you need
+- Change if you need
+- Be flexible
+
++++
+
+Python with C/C++ extensions
+----------------------------
+
+- Python API
+- CPython interface
+- CFFI
+- documented & standardised
+
++++
+
+The ROS way
+-----------
+
+- communicating asynchronous processes
+- common message format
+- compatible serialization/deserialization code
+- basic network management code
+
+
++++
+
+ROS1 Template
+-------------
+
+- A reference for ROS nodes
+- "pure" C++ library
+- C++ / ROS nodes and package
+- ROS / python nodes and package
+- "pure" python nodes and package
+- Nothing beats DIY !
+
+---
+
+ROS C++ Exercises
+=================
+
+- Make a C++ "package" without ROS
+- Alex TODO
+
+---
+
+ROS Python Exercises
+====================
+
+- Make a ROS package
+- Make a Python package
+- Turn it into a ROS package
+- Make a client python script
+- Mutate client into a node
+- Review
+
++++
+
+Make a ROS package
+------------------
+
+- Follow ROS python tutorials
+- Refer to [ros1_pytemplate](https://github.com/pyros-dev/ros1_template/tree/master/ros1_pytemplate) 
+- Implement a Number Node
+- stores a number (initialized at 1)
+- provides reset() for resetting it
+- provides value() for getting its value
+- provides a ROS API for reset() and value()
+
+
++++
+
+Make a Python package
+---------------------
+
+- Follow http://packaging.python.org
+- Refer to [cookiecutters](https://github.com/audreyr/cookiecutter)
+- Provide a python package with Accumulator module
+- Accumulator stores a number N (init at 1) and a constant C
+- Accumulator can do inc_mod(m) -> (N + m) mod C
+
++++
+
+Turn it into a ROS package
+--------------------------
+
+- Use https://github.com/pyros-dev/catkin_pip
+- Refer to [ros1_pip_pytemplate](https://github.com/pyros-dev/ros1_template/tree/master/ros1_pip_pytemplate))
+- Unify tests
+- Make a ROS Accumulator node
+- Choose a ROS API to provide access to inc_mod
+
++++
+
+Make a client python script
+---------------------------
+
+- Not a node
+- Uses Number & Accumulator ROS API
+- Implements a Fibonacci modulo sequence
+- Doesnt do any arithmetic (+, - *, /) itself
+
++++
+
+Mutate client into a node
+-------------------------
+
+- Can we improve the communication
+- Is it really better ?
+- What are the problems or risks ? 
+
++++
+
+Review
```

Comment:
That's the next class already?

Reply:
Yeah, only one presentation for both.
Mmaybe I should add a "break" slide before this one ?